File: frontend/types/repository.ts
export type ImportMethod = "zip" | "git" | "folder";
export type RepositoryStatus = "importing" | "analyzing" | "ready" | "error";
export type ArchitectureStyle =
  | "mvc"
  | "clean"
  | "layered"
  | "microservices"
  | "monolith"
  | "serverless"
  | "unknown";

export interface RepositoryCreate {
  name: string;
  description?: string;
  import_method: ImportMethod;
  source_url?: string;
  source_path?: string;
}

export interface RepositoryInfo {
  id: string;
  name: string;
  description: string | null;
  status: RepositoryStatus;
  import_method: ImportMethod;
  source_url: string | null;
  local_path: string;
  created_at: string;
  analyzed_at: string | null;
  error_message: string | null;
}

export interface LanguageInfo {
  name: string;
  file_count: number;
  total_lines: number;
  percentage: number;
  extensions: string[];
}

export interface FrameworkInfo {
  name: string;
  version: string | null;
  confidence: number;
  evidence: string[];
}

export interface DependencyInfo {
  name: string;
  version: string | null;
  is_production: boolean;
  source_file: string;
}

export interface FolderInfo {
  name: string;
  path: string;
  purpose: string;
  file_count: number;
  children: FolderInfo[];
}

export interface ArchitectureSummary {
  style: ArchitectureStyle;
  description: string;
  entry_points: string[];
  main_modules: string[];
  authentication_detected: boolean;
  database_detected: boolean;
  api_detected: boolean;
  frontend_detected: boolean;
}

export interface CodeElement {
  name: string;
  type: string;
  file_path: string;
  line_start: number;
  line_end: number | null;
  docstring: string | null;
  decorators: string[];
  parent_class: string | null;
  parameters: string[];
  return_type: string | null;
  imports: string[];
  is_exported: boolean;
}

export interface RouteInfo {
  method: string;
  path: string;
  function_name: string;
  file_path: string;
  line_number: number;
  middleware: string[];
  authentication_required: boolean;
  parameters: string[];
  response_model: string | null;
}

export interface DatabaseModelInfo {
  name: string;
  table_name: string | null;
  file_path: string;
  line_number: number;
  fields: CodeElement[];
  relationships: string[];
  primary_key: string | null;
  foreign_keys: string[];
}

export interface AnalysisResult {
  repository_id: string;
  analyzed_at: string;
  languages: LanguageInfo[];
  frameworks: FrameworkInfo[];
  dependencies: DependencyInfo[];
  folder_structure: FolderInfo[];
  architecture: ArchitectureSummary;
  code_elements: CodeElement[];
  routes: RouteInfo[];
  database_models: DatabaseModelInfo[];
  import_graph: any[];
  config_files: string[];
  entry_points: string[];
  total_files: number;
  total_lines: number;
  analysis_time_ms: number;
}

export interface GraphNode {
  id: string;
  type: string;
  name: string;
  metadata: Record<string, any>;
}

export interface GraphEdge {
  source: string;
  target: string;
  relationship: string;
}

export interface SemanticGraph {
  nodes: GraphNode[];
  edges: GraphEdge[];
  root_id: string;
}

File: backend/app/repo_intelligence/schemas/repository.py
"""Repository data models."""

from enum import Enum

from pydantic import BaseModel, Field


class ImportMethod(str, Enum):
    """Supported repository import methods."""

    ZIP = "zip"
    GIT = "git"
    FOLDER = "folder"


class RepositoryStatus(str, Enum):
    """Repository processing status."""

    IMPORTING = "importing"
    ANALYZING = "analyzing"
    READY = "ready"
    ERROR = "error"


class RepositoryCreate(BaseModel):
    """Schema for creating a new repository."""

    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    import_method: ImportMethod
    source_url: str | None = None
    source_path: str | None = None


class RepositoryInfo(BaseModel):
    """Schema for repository information."""

    id: str
    name: str
    description: str | None
    status: RepositoryStatus
    import_method: ImportMethod
    source_url: str | None
    local_path: str
    created_at: str
    analyzed_at: str | None
    error_message: str | None = None


class LanguageInfo(BaseModel):
    """Detected programming language information."""

    name: str
    file_count: int
    total_lines: int
    percentage: float
    extensions: list[str]


class FrameworkInfo(BaseModel):
    """Detected framework information."""

    name: str
    version: str | None
    confidence: float = Field(..., ge=0.0, le=1.0)
    evidence: list[str]


class DependencyInfo(BaseModel):
    """Project dependency information."""

    name: str
    version: str | None
    is_production: bool
    source_file: str


class FolderInfo(BaseModel):
    """Directory structure information."""

    name: str
    path: str
    purpose: str
    file_count: int
    children: list["FolderInfo"] = []


class ArchitectureStyle(str, Enum):
    """Detected architecture style."""

    MVC = "mvc"
    CLEAN = "clean"
    LAYERED = "layered"
    MICROSERVICES = "microservices"
    MONOLITH = "monolith"
    SERVERLESS = "serverless"
    UNKNOWN = "unknown"


class ArchitectureSummary(BaseModel):
    """Architecture analysis summary."""

    style: ArchitectureStyle
    description: str
    entry_points: list[str]
    main_modules: list[str]
    authentication_detected: bool
    database_detected: bool
    api_detected: bool
    frontend_detected: bool
